### ISSUE TITLE
[?] discover_by_ip - Shelly periodically unavailable (very regularly)

### ISSUE THREAD

**Issue by sstefanowski** [StyraHem/ShellyForHASS#272]
I'm testing the usability of "discovery_by_ip" due to my specific network setup described in separate issue here (https://github.com/StyraHem/ShellyForHASS/issues/270). My config setup is also described in the referenced thread.

I noticed that my Shelies (the one which is connected to a weaker "backup" network) goes to the "unavailable" state for less than second but... every minute. The extract from logbook as below:
![image](https://user-images.githubusercontent.com/36473471/79686768-1251ec80-8243-11ea-86d1-b4e51bb22cb9.png)

This creates a log of "noise" in the history graph. This may have a side effect on UI as well, after 1 day of operations, the log contains so much data that it's impossible to render the "history graph" on UI. It takes about a minute to load the data to UI and render the history-bar. Example below
![image](https://user-images.githubusercontent.com/36473471/79686901-fbf86080-8243-11ea-8238-3885210953f6.png)

Additional infos: 
- The RSSI of the Shelly is low (-84dBm) but the switch is still accessible via its Web interface.
- I installed the latest beta 0.1.8.b1 and disabled CoAP, mDns in the configuration by setting "mdns:" and "discovery:" to false. Nothing changes...

Summarizing: This is something strange that plugin reports it's unavailability with such regularity... The regularity suggests this issue is software related (any scheduled operations fail?) not network related (would be more random probably).

**Comment by chemelli74** [comment]
I would again say that is network related to your specific configuration.

The only way to get a realtime and reliable update is to use CoAP, and specifically you need to find "**CoAP-msg**" in your protocols.
If not done yet, see wiki for more info: [https://github.com/StyraHem/ShellyForHASS/wiki/FAQ2:-Slow-updates-of-lovelave-UI](https://github.com/StyraHem/ShellyForHASS/wiki/FAQ2:-Slow-updates-of-lovelave-UI)

Simone

**Comment by sstefanowski** [comment]
I think more-and-more people will have more than one router at home (especially if the house or appartment is big), and having Shellies in different LAN will be quite often case. I've tested many networking setups and this setup (as described in (https://github.com/StyraHem/ShellyForHASS/issues/270)) was up to now best for me - being most reliable and getting fastest transfers on both LANs. My both routers are paired in WISP mode (on 5GHz Wifi), and quite powerful router (Tenda AC6) is broadcasting 2,4Ghz network for my IOT devices.   

What do you actually mean by saying

> ...you need to find "CoAP-msg" in your protocols.

Unfortunately, I have no control over low-level networking options in Tenda AC6. I exposed Home Assistant as "DMZ Host" so theoretically all ports and protocols should be accessible for the devices running in the other network, but somehow CoAP messages from Shelly running in "the other network" are not delivered to HA. 

But isn't it really a bug of the plugin? "Discovery_by_ip" is recommended as an option for setups like mine running different LANs, and it's behaving strange...

Hmmm... it looks like your great plugin (I was really impressed from the very first time I used it) is not an option for me until it has better support for "many LAN setup" - maybe... until it will support fully local MQTT control (presence detection and commands). 




Any other hints?

**Comment by chemelli74** [comment]
I'm sorry is not working for you, but it's all down to a network setup and/or issue.

If you want realtime udpates, only choice is CoAP and we rely on the network delivery of those messages. Nothing we can do from the plugin point of view.

If CoAP is not an option, there are some workarrounds but all of them have pros and contra.

Finally I have a suggestion for you: what about if we add a DNS name support and not only a IP address support ?
That way you can register devices on a local DHCP+DNS server config; when they change address the combo above will update the record and HASS will always find them by the same name.

Would it be a solution ?

If so, I think it can be implemented, as it's not a forced specific logic for your setup, but it's more something that can be widely usefull.

Simone

**Comment by sstefanowski** [comment]
You do not need to be sorry :). You've done really great job. Thank you for your work.

What I just want to say is this plugin could be even more "universal" if it had option to monitor and control local Shellies over MQTT ([Shelly MQTT configs])(https://shelly-api-docs.shelly.cloud/#mqtt-configuration) - Hmmm.... I'm slowly coming to the conclusion that is is the only "universal" solution for most of the multi-LAN setups. MQTT is a TCPIP, which must be supported by all of the routers, and CoAP (based on UDP) may not be supported by some routers.  I know adding MQTT support is probably a lot of work (probably comparable to development of completely new plugin), but I think your plugin extended with MQTT support would be..... a complete solution.

And I know that using CoAP gives nice features like the option to control Shellies from HA and Cloud... but being in your position I would steer the development of plugin to extend where it has the biggest limitations.

**Comment by chemelli74** [comment]
> What I just want to say is this plugin could be even more "universal" if it had option to monitor and control local Shellies over MQTT ([Shelly MQTT configs](https://shelly-api-docs.shelly.cloud/#mqtt-configuration))

This plugin is based on the idea to have both cloud and local access; this won't be possible with MQTT ;-)

Simone

**Comment by chemelli74** [comment]
@sstefanowski, what we do with this issue, based that we won't add MQTT ?

Simone

**Comment by paparapapapapa** [comment]
I second this issue. From 8th of April I have this same problem, with the log flooded by 1 sec unavailabilities. I only saw this today because I was troubleshooting my crashed db!
I don't have a fancy LAN config and I have other 8 shellies-1 that behave normally. I use this custom component only to get attributes more easily, for command I always use MQTT.
MQTT and web interface still work.

Steps I done:
- deleted integration
- uninstalled custom_component
- rebooted HA
- redone all and waited for discovery to rediscover devices
But none changed.

HA 0.108.6 on Docker on NUC.

**Comment by hakana** [comment]
Try to change increase unavailable_after_sec to 70 or decrease scan_interval to 50 or lower.

**Comment by sstefanowski** [comment]
Dear Plugin Team,

I've got also some other information, I was testing for a few days.

In parallel to discussion with you, I opened a ticket with ShellyTeam (#T-20017649) because in my environment the Shelly devices were hanging - disconnecting from Wifi and not reconnecting to any network present. The devices were switchable by the wall switch, but they were not accessible online - I was forced to reboot them by house circuit power/off/on. When hung the devices were not even recording the state - after power reboot, they were going to last remembered state old by a few hours. I was not mentioning this issue here because I suspected this is pure Shelly device issue, but... 

....after I removed ShellyForHass plugin and stayed on pure MQTT integration, my devices seem much more stable - it has been already online two days and no Shelly disconnections. It looks like a) the Shelly firmware has a bug when under some conditions it may hang b) the plugin uses it in a way it may cause device hanging - maybe this is intensive use of API (in my setup) or maybe it's because "discovery_by_ip" option?

So.... my local configuration when the problem was really frequent was:
- all of my Shellies under latest firmware - 20200320-123430/v1.6.2@514044b4
- ShellyCloud turned off in devices - local MQTT service registered
- ShellyForHass plugin installed with CoAP discovery, later with igmp_fix - and later also with "discovery_by_ip" option - all describe in this thread.

I hope you can provide to the Shelly team some information helping them improving Shelly firmware.

**Comment by chemelli74** [comment]
I already saw issues when both MQTT and CoAP were used.
I would ask a test with only CoAP to double check your findings.

Also, if CoAP is not CoAP-msg, like for you, I suspect a memory leak in Shelly's firmware due to many connections to REST api.

Let us know your findings.

Simone

**Comment by sstefanowski** [comment]
Note: I do not want to connect to Shelly Cloud.
How would you like me to test it? Do you suggest I should just disable MQTT in Shelly devices and try your plugin again? Do I understand right?

**Comment by hakana** [comment]
Yes, please test to disable MQTT and see if that help.

**Comment by chemelli74** [comment]
@sstefanowski , any news ?

Simone

**Comment by paparapapapapa** [comment]
Ciao Simone, after some investigation I found it was a network fault, I had to throw away one of my AP's. I didn't realize in that moment that said AP served only that particular Shelly. All other Shellies are working fine as usual.
Grazie anyway.
R.

**Comment by sstefanowski** [comment]
Hi Simone,
I'm still testing various configurations- it takes some time. Just give me some more time... I'll promise to let you know...

**Comment by jeepotvin** [comment]
Hi. I have two devices connected with HA through this integration. Switches and energy consumption reporting works great. I've only been using my shellies and the integration for the last week or so but since the beginning, I also noticed a lot of **device unavailable** in the logbook. This specific device is 6 feet away from the main Wi-Fi router so signal strength is not an issue.

![2020-04-30_15-22-36](https://user-images.githubusercontent.com/6979157/80750944-f548d380-8af6-11ea-8a17-2b70698f2144.png)

Both shellies have a static IP address provided by the router but haven't use the static IP feature in the shelly firmware yet.

I've read the replies on this issue and will try previous suggestions such as disabling MQTT and maybe play around with other settings like discovery or discover_by_ip. I'll report back here as soon as tested.

**Comment by chemelli74** [comment]
Please update to latest beta (0.1.8.b2) and report back.

Simone

**Comment by jeepotvin** [comment]
Updated to 0.1.8.b2 as requested yesterday night. Rebooted HA.
As of now, it seems to do the trick as there a lot less _unavailable_ in the logbook.

![chrome_2020-05-03_13-07-18](https://user-images.githubusercontent.com/6979157/80920634-11d74c80-8d3f-11ea-8b9f-5cc1958116a3.png)

Note that it isn't the same Shelly device in the screenshot but both had the problem prior to the update. Screenshot taken at 1PM so big improvements regarding logs.

**Comment by hakana** [comment]
Can we close this issue?

**Comment by jeepotvin** [comment]
I'm not the reporter of this issue but it definitely fixed it for me. No data since updating to 0.1.8.b2 or 0.1.8 reveals _unavailable_ for both my Shellys.

![2020-05-07_21-50-54](https://user-images.githubusercontent.com/6979157/81361679-0cfef980-90ad-11ea-8973-18b3ca4276df.png)

**Comment by chemelli74** [comment]
> Hi Simone,
> 
> I'm still testing various configurations- it takes some time. Just give me some more time... I promise, I'll let you know...

We miss just your feedback before closing this issue.
How are your tests going ?

Simone

**Comment by sstefanowski** [comment]
Well, I'm still testing.... The tests are taking long time because myShellies are hanging usually after a few days. 

I just got another issue (one of he shellies hung), and it.... shows that most probably the issue is between, one of "Tenda router" and Shellies (possibly memory leak in Shelly firmware). 
I'm suspecting networking issue as when I connected a few Shellies to different router - even with very weak signal they still work OK (it usually hung fequently when it was connected to Tenda). But I don't suspect the Tenda issue but Shelly firmware issue becasue no other devices had any problem with Tenda so far.
Until now....  I was still testing with MQTT on. I'm planning to disable MQTT completly and just rely on CoAPs and see it that changes anything between Tenda and Shelly. In other words I was to check if Shelly issue is related to Shelly-MQTT-stack or any other Shelly-API.

Now...
If you consider the issue as purely no-plugin-related but Shelly-related you can close the ticket. 
If you want to wait for my feedback (just to provide some information to community Shelly may hung case in strange networking conditions not necessarily related to plugin) please keep this ticket open and I'll update it again in a few days.

**Comment by chemelli74** [comment]
> If you want to wait for my feedback (just to provide some information to community that Shelly may hung case in strange networking conditions not necessarily related to plugin) please keep this ticket open and I'll update it again in a few days.

Thx for your reports, no problem in waiting a few days more.

Simone

**Comment by chemelli74** [comment]
@sstefanowski , did you manage to complete your tests ? I'm curious of your findings.

Simone

**Comment by chemelli74** [comment]
No update for 21 days, closing

Simone
